Assets/Mirror/Runtime/NetworkWriter.cs

```diff
@@ -33,6 +33,21 @@ public byte[] ToArray()
             return ((MemoryStream)writer.BaseStream).ToArray();
         }
 
+        // Gets the serialized data in an ArraySegment<byte>
+        // this is similar to ToArray(),  but it gets the data in O(1)
+        // and without allocations.
+        // Do not write anything else or modify the NetworkWriter
+        // while you are using the ArraySegment
+        public ArraySegment<byte> ToArraySegment()
+        {
+            writer.Flush();
+            if (((MemoryStream)writer.BaseStream).TryGetBuffer(out ArraySegment<byte> data))
```

Comment:
this will never happen at the moment.
@paulpach we will need to create the memorystream with this constructor:
```csharp
(byte[] buffer, int index, int count, bool writable, bool publiclyVisible)
```

see MemoryStream source: https://referencesource.microsoft.com/#mscorlib/system/io/memorystream.cs,9675195c73865c6a

Reply:
Why?

Reply:
@MichalPetryka see the link. internally the exposed flag is only set with the capacity constructor and the constructor that I quoted

Reply:
@vis2k but with the constructor we use it sets it to true

Reply:
I just tested it by adding an exception and it works as is.  No exception

Reply:
true, nevermind paul

Reply:
@vis2k  this is the source of MemoryStream:

```cs
 public MemoryStream() 
            : this(0) {
        }
        
        public MemoryStream(int capacity) {
            if (capacity < 0) {
                throw new ArgumentOutOfRangeException("capacity", Environment.GetResourceString("ArgumentOutOfRange_NegativeCapacity"));
            }
            Contract.EndContractBlock();
 
            _buffer = new byte[capacity];
            _capacity = capacity;
            _expandable = true;
            _writable = true;
            _exposable = true;
            _origin = 0;      // Must be 0 for byte[]'s created by MemoryStream
            _isOpen = true;
        }

...
 public virtual bool TryGetBuffer(out ArraySegment<byte> buffer) {
            if (!_exposable) {
                buffer = default(ArraySegment<byte>);
                return false;
            }
 
            buffer = new ArraySegment<byte>(_buffer, offset:_origin, count:(_length - _origin));
            return true;
        }
 ```

As you can see MemoryStreams created with the default constructor are exposable.  I added an exception there to make sure,  and it is working fine